website/addons/gitlab/views.py

```diff
@@ -0,0 +1,444 @@
+            return None
+
+    try:
+        branch, sha, branches = get_refs(node_settings, branch=kwargs.get('branch'), sha=kwargs.get('sha'), connection=connection)
+    except (NotFoundError, GitLabError):
+        logger.error('GitLab repo not found')
+        return
+
+    if branch is not None:
+        ref = ref_to_params(branch, sha)
+        can_edit = check_permissions(node_settings, auth, connection, branch, sha, repo=repo)
+    else:
+        ref = None
+        can_edit = False
+
+    permissions = {
+        'edit': can_edit,
+        'view': True,
+        'private': node_settings.is_private
+    }
+    urls = {
+        'upload': node_settings.owner.api_url + 'gitlab/file/' + (ref or ''),
+        'fetch': node_settings.owner.api_url + 'gitlab/hgrid/' + (ref or ''),
+        'branch': node_settings.owner.api_url + 'gitlab/hgrid/root/',
+        'zip': node_settings.owner.api_url + 'gitlab/zipball/',
+        'repo': 'https://{0}/{1}/tree/{2}'.format(node_settings.external_account.display_name, node_settings.repo, branch)
+    }
+
+    branch_names = [each['name'] for each in branches]
+    if not branch_names:
+        branch_names = [branch]  # if repo un-init-ed then still add default branch to list of branches
+
+    return [rubeus.build_addon_root(
+        node_settings,
+        repo['path_with_namespace'],
+        urls=urls,
+        permissions=permissions,
+        branches=branch_names,
+        private_key=kwargs.get('view_only', None),
+        defaultBranch=repo['default_branch'],
+    )]
+
+#########
+# Repos #
+#########
+
+@must_have_addon(SHORT_NAME, 'user')
+@must_have_addon(SHORT_NAME, 'node')
+@must_be_addon_authorizer(SHORT_NAME)
+@must_have_permission('write')
+def gitlab_create_repo(**kwargs):
+    repo_name = request.json.get('name')
+    user = request.json.get('user')
+
+    if not repo_name:
+        raise HTTPError(http.BAD_REQUEST)
+
+    node_settings = kwargs['node_addon']
+    connection = GitLabClient(external_account=node_settings.external_account)
+
+    try:
+        repo = connection.create_repo(repo_name, auto_init=True)
+    except GitLabError:
+        raise HTTPError(http.BAD_REQUEST)
+
+    return {
+        'user': user,
+        'repo': repo['path_with_namespace'],
+    }
+
+def add_hook_log(node, gitlab, action, path, date, committer, include_urls=False,
+                 sha=None, save=False):
+    """Add log event for commit from webhook payload.
+
+    :param node: Node to add logs to
+    :param gitlab: GitLab node settings record
+    :param path: Path to file
+    :param date: Date of commit
+    :param committer: Committer name
+    :param include_urls: Include URLs in `params`
+    :param sha: SHA of updated file
+    :param save: Save changes
+
+    """
+    gitlab_data = {
+        'user': gitlab.user,
+        'repo': gitlab.repo,
+    }
+
+    urls = {}
+
+    if include_urls:
+        # TODO: Move to helper function
```

Comment:
Is this necessary? It looks like the only other place it's happening is in a test, so I'd probably say it's not.